### ISSUE TITLE
Problems with the unification algorithm

### ISSUE THREAD

**Issue by schani** [quicktype/quicktype#255]
We would ideally want to unify all types that will end up as unified in one step.  Consider this case:

    [ { “foos”: [ A, B ] },
      { “foos”: [ C, D, E ] } ]

In our current unification algorithm we first unify `A` and `B` giving some type `AB`, then we unify `C` and `D` giving `CD`, which we unify with `E`, giving `CDE` and then we finally unify `AB` with `CDE`.

Let’s say we wanted to detect enums.  If we have five string values, all five of which are different, we should probably guess that the type is a string.  But if we have five thousand strings, but only five distinct values, we should probably guess an enum.  Our current algorithm doesn’t allow us to see all those five thousand values all at once.  A stopgap measure might be to add counts to different values in some cases, like strings, to be able to detect enums, but I think that breaks down very easily.

Another case we’ll want to be able to handle eventually is discriminated unions, where there is one enum property for example that determines what the other properties are. I think we’d also want to be able to look at all types at the same time to figure out whether a discriminated union applies.

The algorithmic difference is that we need to do type generation top-down, as opposed to bottom-up, and have the unification function take any number of values (not types?), as opposed to two types to be unified.  So in the example above we would start by generating a type for the array of the two objects.  That would trigger the type generation for their `foos` properties, which means generating a type for the two arrays.  As long as we don’t have tuples, to generate a type for a bunch of arrays means generating a type for the elements of ALL those arrays combined, i.e for `A`, `B`, `C`, `D`, and `E`.

To detect discriminated unions when looking at a bunch of objects, we could start by checking whether they have any properties in common that are strings (or ints?).  Those would be candidates for the discriminating enum.  Then check whether any of them discriminate sufficiently between the rest of the properties. We could detect subtypes in a similar way: If the number of distinct sets of properties that a bunch of objects have is sufficiently small, they can be represented by different class types. If there exist subset relationships between some of those sets, those can be represented by subtypes.

**Comment by schani** [comment]
Doing top-down type unification requires having the whole JSON object (or objects) in memory.  We can save a lot of memory by leaving out details we don't need and encoding more efficiently.  In particular, these are details we don't need (yet) to infer types:

* The values of numbers.
* The values of strings that we wouldn't consider for enums.  Property names are exempt - we always need those.

I propose representing JSON values by small integers.  We can use a few bits (4 should be enough) as a tag to represent the type, and the remaining bits as an index into a table for non-primitive types.  In particular:

* We maintain an array of strings which the string type indexes into.
* Arrays are stored as arrays in an array of arrays.
* Objects are stored as objects in an array of objects.  We might find that it's more efficient to represent objects as arrays that contain keys and values interleaved, but for a first implementation that's not necessary.

If we have a streaming JSON reader we can convert on the fly into this efficient format.

**Comment by schani** [comment]
Actually, we shouldn't store objects as objects, because that doesn't preserve the order of properties, which is something we've been wanting to do for a while.

I looked at streaming JSON parsers.  I found [clarinet](https://github.com/dscape/clarinet) and [stream-json](https://github.com/uhop/stream-json), which we already use.  I ran a quick benchmark on a large file, and `stream-json` is faster, and it also has more features.
